### PR TITLE
Add Ancestor App IDs Retrieval and Child App IDs Retrieval Support for a Given Shared App

### PR DESCRIPTION
**Pull request by dhaura**
## Purpose
- Implement following methods.
    - `getAncestorAppIds` - Get shared ancestor application IDs for the given shared application ID and and its organization ID.
    - `getChildAppIds` - Get shared child application IDs of the given shared parent application, filtered using the given child organization IDs. 

## Related PRs
- https://github.com/wso2/identity-organization-management-core/pull/141

## Related Issues
- https://github.com/wso2/product-is/issues/20616

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+import java.util.List;
+import java.util.Map;
+
+import javax.sql.DataSource;
+
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.when;
+
+/**
+ * Unit tests for OrgApplicationMgtDAOImpl..
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java:47]:
```suggestion
 * Unit tests for OrgApplicationMgtDAOImpl.
```

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    private static final String PRIMARY_USERSTORE = "PRIMARY";
+    private static final String ADMIN_USERNAME = "admin";
+    private static final String AUTH_TYPE = "default";
+
+    private static final String CREATE_APPLICATION_SQL_STATEMENT = "INSERT INTO SP_APP " +
+            "(TENANT_ID, APP_NAME, USER_STORE, USERNAME, AUTH_TYPE, UUID) VALUES (?, ?, ?, ?, ?, ?);";
+    private static final String SHARE_APPLICATION_SQL_STATEMENT = "INSERT INTO SP_SHARED_APP " +
+            "(MAIN_APP_ID, OWNER_ORG_ID, SHARED_APP_ID, SHARED_ORG_ID, SHARE_WITH_ALL_CHILDREN) " +
+            "VALUES (?, ?, ?, ?, ?);";
+    private static final String DELETE_APPLICATION_SQL_STATEMENT = "DELETE  FROM SP_APP " +
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java:77]:
```suggestion
    private static final String DELETE_APPLICATION_SQL_STATEMENT = "DELETE FROM SP_APP " +
```

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            "WHERE UUID = ?;";
+    private static final String DELETE_SHARED_APPLICATION_SQL_STATEMENT = "DELETE  FROM SP_SHARED_APP " +
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java:79]:
```suggestion
    private static final String DELETE_SHARED_APPLICATION_SQL_STATEMENT = "DELETE FROM SP_SHARED_APP " +
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java [hunk 4]
```diff
@@ -118,4 +119,22 @@ Optional<String> getSharedApplicationResourceId(String mainAppId, String ownerOr
...
             throws OrganizationManagementException;
+
+    /**
+     * Returns the unique identifiers of shared applications associated with the given main application
+     * within the given shared organizations.
+     *
+     * @param mainAppId    The app ID of the main application.
+     * @param ownerOrgId   The organization ID of the owner.
+     * @param sharedOrgIds The list of app shared organization IDs.
+     * @throws OrganizationManagementException The server exception is thrown in a failure
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java:130]: returned item is not described in the comment

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 5]
```diff
@@ -902,6 +968,23 @@ private boolean shouldUpdateShareWithAllChildren(boolean shareWithAllChildren, S
         return false;
     }
 
+    /**
+     * Filter and return shared child applications of the given main application for the given child organization IDs.
+     *
+     * @param mainAppId Application ID of the main application.
+     * @param mainOrgId Organization ID of the main application.
+     * @return The map containing organization ID and application ID of the filtered shared applications.
+     */
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:977]: exception is not in doc comment

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java (hunk 1 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/6bc8b6f5e9181faffb2c20befe3986561c7bce75

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java (hunk 2 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/6bc8b6f5e9181faffb2c20befe3986561c7bce75

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java (hunk 3 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/dao/impl/OrgApplicationMgtDAOImplTest.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/6bc8b6f5e9181faffb2c20befe3986561c7bce75

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java (hunk 4 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java:131]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/6bc8b6f5e9181faffb2c20befe3986561c7bce75

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java (hunk 5 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:978]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/6bc8b6f5e9181faffb2c20befe3986561c7bce75

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 6]
```diff
@@ -143,5 +145,35 @@ default String getMainApplicationIdForGivenSharedApp(String sharedAppId, String
...
+            throws OrganizationManagementException {
+
+        throw new NotImplementedException(
+                "getAncestorAppIds method is not implemented in " + this.getClass().getName());
+    }
+
+    /**
+     * Get shared child application IDs of the given parent application.
+     *
+     * @param parentAppId The unique ID of the shared parent application.
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:168]:
```suggestion
     * @param parentAppId The unique ID of the root app / shared app in parent org.
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 7]
```diff
@@ -143,5 +145,35 @@ default String getMainApplicationIdForGivenSharedApp(String sharedAppId, String
...
+     * @param parentOrgId The organization ID of the parent.
+     * @param childOrgIds The organization ID list of the children.
+     * @return The map containing organization ID and application ID of the shared child applications.
+     * @throws OrganizationManagementException If errors occurred when retrieving the child app IDs.
+     */
+    default Map<String, String> getChildAppIds(String parentAppId, String parentOrgId, List<String> childOrgIds)
+            throws OrganizationManagementException {
+
+        throw new NotImplementedException("getParentAppId method is not implemented in " + this.getClass().getName());
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:177]:
```suggestion
        throw new NotImplementedException("getChildAppIds method is not implemented in " + this.getClass().getName());
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 8]
```diff
@@ -655,6 +656,93 @@ public void shareApplication(String ownerOrgId, String sharedOrgId, ServiceProvi
...
+    /**
+     * Returns whether the given application is a main application.
+     *
+     * @param appId The unique ID of the application.
+     * @param orgId The organization ID of the given application.
+     * @return Returns true if the given application is a main application.
+     * @throws OrganizationManagementException If an error occurs while retrieving the application.
+     */
+    private boolean isMainApp(String appId, String orgId)
+            throws OrganizationManagementException {
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:720]: Can't we move this to above line?

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 9]
```diff
@@ -655,6 +656,93 @@ public void shareApplication(String ownerOrgId, String sharedOrgId, ServiceProvi
...
+
+        OrganizationManager organizationManager = OrgApplicationMgtDataHolder.getInstance().getOrganizationManager();
+        String parentTenantDomain = organizationManager.resolveTenantDomain(orgId);
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:723]:
```suggestion
        String tenantDomain = organizationManager.resolveTenantDomain(orgId);
```

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImplTest.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+import java.util.stream.Collectors;
+
+import static org.mockito.ArgumentMatchers.anyList;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.wso2.carbon.identity.organization.management.application.constant.OrgApplicationMgtConstants.IS_FRAGMENT_APP;
+
+/**
+ * Unit tests for OrgApplicationManagerImpl..
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImplTest.java:51]:
```suggestion
 * Unit tests for OrgApplicationManagerImpl.
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 11]
```diff
@@ -143,5 +145,35 @@ default String getMainApplicationIdForGivenSharedApp(String sharedAppId, String
...
-            boolean shareWithAllChildren) throws OrganizationManagementException;
+                          boolean shareWithAllChildren) throws OrganizationManagementException;
+
+    /**
+     * Get the shared ancestor application IDs for the given child application ID of the given child organization.
+     *
+     * @param childAppId The unique ID of the shared child application.
+     * @param childOrgId The organization ID of the child.
+     * @return Map containing shared ancestor application IDs and their organization IDs.
+     * @throws OrganizationManagementException If errors occurred when retrieving the parent application ID.
```
[the remark this answers is not in the document]
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/4a880cda919340753b6c7bfbad302ced4aef7fcd

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java (hunk 6 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/4a880cda919340753b6c7bfbad302ced4aef7fcd

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java (hunk 7 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/4a880cda919340753b6c7bfbad302ced4aef7fcd

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java (hunk 8 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/4a880cda919340753b6c7bfbad302ced4aef7fcd

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java (hunk 9 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/4a880cda919340753b6c7bfbad302ced4aef7fcd

File: components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImplTest.java (hunk 10 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.organization.management.application/src/test/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImplTest.java]: Addressed with https://github.com/wso2-extensions/identity-organization-management/pull/383/commits/4a880cda919340753b6c7bfbad302ced4aef7fcd

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**